esp-hal/src/system.rs

```diff
@@ -229,30 +229,33 @@ impl<const NUM: u8> crate::peripheral::Peripheral for SoftwareInterrupt<NUM> {
 impl<const NUM: u8> crate::private::Sealed for SoftwareInterrupt<NUM> {}
 
 /// This gives access to the available software interrupts.
-///
-/// Please note: Software interrupt 0 is not available when using the
-/// `embassy-executor-thread` feature
+#[cfg_attr(
+    multi_core,
+    doc = r#"
+Please note: Software interrupt 3 is reserved
+for inter-processor communication when the `embassy`
+feature is enabled."#
+)]
 #[non_exhaustive]
 pub struct SoftwareInterruptControl {
-    #[cfg(not(all(feature = "embassy-executor-thread", multi_core)))]
     pub software_interrupt0: SoftwareInterrupt<0>,
     pub software_interrupt1: SoftwareInterrupt<1>,
     pub software_interrupt2: SoftwareInterrupt<2>,
+    #[cfg(not(all(feature = "embassy", multi_core)))]
     pub software_interrupt3: SoftwareInterrupt<3>,
 }
 
 impl SoftwareInterruptControl {
     fn new() -> Self {
-        // the thread-executor uses SW-INT0 when used on a multi-core system
-        // we cannot easily require `software_interrupt0` there since it's created
-        // before `main` via proc-macro
-
         SoftwareInterruptControl {
-            #[cfg(not(all(feature = "embassy-executor-thread", multi_core)))]
-            software_interrupt0: SoftwareInterrupt,
-            software_interrupt1: SoftwareInterrupt,
-            software_interrupt2: SoftwareInterrupt,
-            software_interrupt3: SoftwareInterrupt,
+            software_interrupt0: SoftwareInterrupt {},
+            software_interrupt1: SoftwareInterrupt {},
+            software_interrupt2: SoftwareInterrupt {},
+            // the thread-executor uses SW-INT3 when used on a multi-core system
```

Comment:
currently esp-wifi uses FROM_CPU3 for yielding ... we need to change it there (and the initialize function should probably take it if it doesn't already)

Reply:
It might actually be nice to add some comments relating to this as well. For instance I have no idea why it was 0 before and is 3 now, or whether or not this number matters or what the potential repercussions are.

Reply:
No repercussions (just that we will need to modify esp-wifi a bit) that I'm aware of, it was a personal choice that cfg away the last software interrupt instead of the first one. It makes more sense to me to do it this way instead of the software interrupts available to users start at 1.

I could revert this change, I'm not that attached to it.

Reply:
Sorry! I should have put that in the PR desc, this PR kinda got out of hand with the number of changes compared to what I initially intended :D.